pkg/agent/openflow/pipeline.go

```diff
@@ -2575,6 +2590,62 @@ func (c *client) createOFTable(table binding.Table, nextID uint8, missAction bin
 	return table
 }
 
+// igmpPktInFlows sends the IGMP packets to Antrea Agent on L3ForwardingTable.
+func (c *client) igmpPktInFlows(reason uint8) []binding.Flow {
+	pktInTable := L3ForwardingTable
+	flows := []binding.Flow{
+		// Send the IGMP packet to Antrea Agent and forward the packet normally in the OVS bridge if the packet is set
+		// with the "CustomReasonIGMPRegMark", so that the OVS multicast db cache can be updated, and Antrea Agent can
+		// identify the local multicast group and its members in the meanwhile.
+		pktInTable.BuildFlow(priorityHigh).
+			MatchProtocol(binding.ProtocolIGMP).
+			MatchRegMark(CustomReasonIGMPRegMark).
+			Action().SendToController(reason).
+			Action().Normal().Done(),
+	}
+	for _, dstIP := range []net.IP{mcastAllRouters, igmpV3ReportDst} {
+		// Set a custom reason for the IGMP packets in MulticastTable and send it to the pktInTable. "pktInTable" is
+		// only used to packet in the IGMP packet, and the other multicast traffic is forwarded normally or output to
+		// antrea-gw0 in the MulticastTable by the flows installed by "localMulticastForwardFlow" or
+		// "externalMulticastReceiverFlow".
+		flows = append(flows, MulticastTable.BuildFlow(priorityHigh).
+			MatchProtocol(binding.ProtocolIGMP).
+			MatchRegMark(FromLocalRegMark).
+			MatchDstIP(dstIP).
+			Action().LoadRegMark(CustomReasonIGMPRegMark).
+			Action().GotoTable(pktInTable.GetID()).
+			Done())
+	}
+	return flows
+}
+
+// localMulticastForwardFlow forwards the multicast traffic with OVS action "normal", and outputs it to antrea-gw0 in
+// the meanwhile. So that the packet could be forwarded to local Pods which have joined the Multicast group and to the
+// external receivers. For the external multicast traffic accessing to the given multicastIP also hits this flow, and
+// the packet is not sent back to antrea-gw0 because OVS datapath will drop it when it finds the output port is the same
+// as the input port.
+func (c *client) localMulticastForwardFlow(multicastIP net.IP) []binding.Flow {
+	return []binding.Flow{
+		MulticastTable.BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchDstIP(multicastIP).
+			Action().Output(config.HostGatewayOFPort).
+			Action().Normal().
+			Done(),
+	}
+}
+
+// externalMulticastReceiverFlow outputs the multicast traffic to antrea-gw0, so that local Pods can send multicast traffic
+// to access the external receivers. For the case that one or more local Pods have joined the target multicast group,
+// it is handled by the flows created by function "localMulticastForwardFlow" after local Pods report the IGMP membership.
+func (c *client) externalMulticastReceiverFlow() binding.Flow {
+	return MulticastTable.BuildFlow(priorityLow).
+		MatchProtocol(binding.ProtocolIP).
+		MatchDstIPNet(*mcastCIDR).
+		Action().Output(config.HostGatewayOFPort).
```

Comment:
I'm trying to understand why it needs two kinds of flows for groups that have been discovered and groups that haven't been discovered. If we are going to forward it to external network anyway and `Normal` can handle the multicast automatically based on its cache. Why don't we just use a flow that matches `mcastCIDR` and has  `Output(config.HostGatewayOFPort)` and `Normal` actions?